pkg/apis/pipeline/v1alpha1/resource_types.go

```diff
@@ -164,3 +164,19 @@ func ResourceFromType(r *PipelineResource) (PipelineResourceInterface, error) {
 	}
 	return nil, fmt.Errorf("%s is an invalid or unimplemented PipelineResource", r.Spec.Type)
 }
+
+// AttributesFromType returns a list of available attributes that can be replaced
+// through templating.
+func AttributesFromType(prt PipelineResourceType) ([]string, error) {
```

Comment:
we could have some unit tests for `AttributesFromType` too :D

Reply:
Glad you recommended this! Ended up running into an issue where this function was not working for one of the types (`PipelineResourceTypeStorage`). I included a fairly kludgy solution and a comment to help illustrate the problem.

> We have run into this unfortunate event where technically we cannot determine what attributes an object has strictly from the PipelineResourceType alone. This is because of the subtypes that exists for the PipelineResourceTypeStorage type. At the moment, both of the subtypes happen to have equivalent keys in their Replacements(). So technically we can just pick one of the subtypes and still return the proper attributes. However, we may want to reconsider the approach on how we obtain or store these attributes.

I do not have more time to look at this at the moment, but will find time soon to try to come up with a more elegant solution.